### ISSUE TITLE
Allow users to adjust HTTP headers (Authentication, User-Agent, Accept) when creating projects from Web Addresses (URLs)

### ISSUE THREAD

**Issue by DavidFichtmueller** [OpenRefine/OpenRefine#2104]
**Is your feature request related to a problem or area of OpenRefine? Please describe.**
When creating a project by getting the data from "Web Addresses (URLs)" it is currently not possible to change the nature of the request by altering the HTTP headers for Authentication, User-Agent, Accept. 
![OpenRefine_Get_data_from_Web_Addresses](https://user-images.githubusercontent.com/7805030/62195137-a6111f00-b37b-11e9-8601-37be43102219.PNG)

If there were such an option, it would be possible to load datasets directly from remote protected APIs or request data in a format that can be ingested by OpenRefine, instead of having to download it manually and importing the file.

**Describe the solution you'd like**
In #218 the support of custom HTTP headers was added to the "add column by fetching URLs" dialog. The same fields should be available when loading a dataset directly from an URL.  
![Open_Refine_Add_column_by_fetching_URLs](https://user-images.githubusercontent.com/7805030/62195160-af01f080-b37b-11e9-9588-f816e5e681fc.PNG)

**Describe alternatives you've considered**
The alternative would be to download the datasets manually, store them as file and import them as "Get data from ... This Computer". For my specific usecase described below, it would also be possible to set up a proxy server that adds the Accept header to the request and returns the result.

**Additional context**
My specific use case for this feature is the possibility to load data directly from the Wikidata Query Service. For instance [this query](https://query.wikidata.org/sparql?query=SELECT%20DISTINCT%20%3Fcity%20%3FcityLabel%20%3Fcountry%20%3FcountryLabel%20%3Flocation%20%3Fpopulation%20%3Farea%20%3Finception%20%3Fisocode%20%3Felevation%0AWHERE%20%0A%7B%0A%20%20%3Fcity%20p%3AP31%2Fwdt%3AP279*%20%3Fstatement.%0A%20%20%3Fstatement%20ps%3AP31%20wd%3AQ5119.%0A%20%20FILTER%20NOT%20EXISTS%20%7B%20%3Fstatement%20pq%3AP582%20%3Fx%20%7D%0A%20%20%3Fcity%20wdt%3AP1376%20%3Fcountry.%0A%20%20%3Fcity%20wdt%3AP17%20%3Fcountry.%0A%20%20%3Fcity%20wdt%3AP625%20%3Flocation.%0A%20%20%3Fcity%20wdt%3AP1082%20%3Fpopulation.%0A%20%20%3Fcity%20wdt%3AP2046%20%3Farea.%0A%20%20OPTIONAL%7B%3Fcity%20wdt%3AP571%20%3Finception.%7D%0A%20%20OPTIONAL%7B%3Fcity%20wdt%3AP300%20%3Fisocode.%7D%0A%20%20OPTIONAL%7B%3Fcity%20wdt%3AP2044%20%3Felevation.%7D%0A%20%20SERVICE%20wikibase%3Alabel%20%7B%20bd%3AserviceParam%20wikibase%3Alanguage%20%22en%2Cen%22.%20%7D%0A%7DORDER%20BY%20DESC(%3Fpopulation) ) retrieves information about the worlds capitols. When requested as such in the browser (or via OpenRefine) it delivers the result set as XML. But when the "Accept" header is set to "text/csv" the data comes in a format the can be directly imported by OpenRefine. With the ever increasing interconnectedness between OpenRefine and Wikidata, I think that many others would profit from such a change as well.

**Comment by wetneb** [comment]
Thanks for this flawless ticket, it is a pleasure to read! I think what you are requesting makes perfect sense, it seems reasonable to give control of such options to the user at this stage. There is plenty of space in this UI to add these options, so it's just a matter of tweaking that UI and adding the corresponding support in the backend.

Related: #1212, which would let you type your SPARQL query directly in OpenRefine.

If someone wants to have a go at this, that UI is defined here:
https://github.com/OpenRefine/OpenRefine/blob/master/main/webapp/modules/core/scripts/index/default-importing-sources/sources.js#L74 and
https://github.com/OpenRefine/OpenRefine/blob/master/main/webapp/modules/core/scripts/index/default-importing-sources/import-from-web-form.html

In the backend, the corresponding place where the download is actually done is
https://github.com/OpenRefine/OpenRefine/blob/master/main/src/com/google/refine/importing/ImportingUtilities.java

**Comment by ostephens** [comment]
Note also that that the headers supported by the "add column by fetching URLs" function are added via https://github.com/OpenRefine/OpenRefine/blob/master/main/src/com/google/refine/commands/HttpHeadersSupport.java and the way they are registered should make them available throughout the system

This list of headers is retrieved by the GetModels command https://github.com/OpenRefine/OpenRefine/blob/master/main/src/com/google/refine/commands/project/GetModelsCommand.java

In the UI there is an example of this data being read out in https://github.com/OpenRefine/OpenRefine/blob/master/main/webapp/modules/core/scripts/dialogs/http-headers-dialog.js

I'd hope that much of this is re-usable so there is a single place where supported headers are registered and this would then be used in both places (import data from web and add data via URL)

**Comment by tfmorris** [comment]
I agree that HTTP connections should be handled uniformly for Create Project and Fetch URL (and Reconciliation for that matter), but I'd like to see the hand specified Authorization header removed. This should be managed the same way a browser would do it by responding to the 401 error and challenge to prompt the user for basic auth info.

**Comment by sjathin** [comment]
@ostephens - I see that in ```HttpHeadersDialog```, the headers are fetched using ```httpHeaders``` of ```theProject```.  But ```theProject``` isn't available at this stage(project isn't created yet). Any suggestions as to how I can reuse this?

**Comment by ostephens** [comment]
@sjathin I'm not entirely sure of the best approach, but my first thought is to add a `get-httpheaders` command following the pattern of `get-languages`.  I think this would be:

* Add a `GetHttpHeadersCommand` class (similar to GetLanguagesCommand) which returns the list of headers & default values as JSON
* Register the command in controller.js
* Add Ajax call to this new command to index.js

@wetneb or @tfmorris might have other suggestions for the best approach

**Comment by sjathin** [comment]
> Thanks for this flawless ticket, it is a pleasure to read! I think what you are requesting makes perfect sense, it seems reasonable to give control of such options to the user at this stage. There is plenty of space in this UI to add these options, so it's just a matter of tweaking that UI and adding the corresponding support in the backend.
> 
> Related: #1212, which would let you type your SPARQL query directly in OpenRefine.
> 
> If someone wants to have a go at this, that UI is defined here:
> https://github.com/OpenRefine/OpenRefine/blob/master/main/webapp/modules/core/scripts/index/default-importing-sources/sources.js#L74 and
> https://github.com/OpenRefine/OpenRefine/blob/master/main/webapp/modules/core/scripts/index/default-importing-sources/import-from-web-form.html
> 
> In the backend, the corresponding place where the download is actually done is
> https://github.com/OpenRefine/OpenRefine/blob/master/main/src/com/google/refine/importing/ImportingUtilities.java

@wetneb - should "Accept" header be set to "text/csv"(form data) for the request in which the sub command is ```load-raw-data``` ? But I see in the network that the data is downloaded for the request with sub command - ```default-importing-controller``` ?
